### PR TITLE
Update license files

### PR DESCRIPTION
**Pull request by TedThemistokleous**
Updating license at the top of AMD developed files. Created a script to achieve this called license_stamper.py 

All this does is pull in what LICENSE says, and adds that with the proper delimiters on top of our files. Let me know if I should be adding additional filenames. This is based on using what files are tracked in git, then running through them to stamp things.

I've also included a git hook to keep the date constantly rolling found from: https://github.com/ROCmSoftwarePlatform/rocBLAS-internal/pull/1214/files#

I can remove the above if it leads to any issues.

### CODE REVIEW

File: .githooks/pre-commit [hunk 1]
```diff
@@ -45,6 +45,18 @@ else
     against=16bbb57
 fi
 
+[[ -z "$files" ]] && exit
+# Change the copyright date at the top of any text files
```
Comment (pfultz2) [inline .githooks/pre-commit:49]: Does this apply to all files or just the ones that were changed in the commit?

File: .githooks/pre-commit [hunk 2]
```diff
@@ -45,6 +45,18 @@ else
...
+for file in $files; do
+    [[ -L $file ]] && continue
+    echo "Processing copyright dates in $file"
+    if [[ -e $file ]]; then
+        /usr/bin/perl -pi -e 'INIT { exit 1 if !-f $ARGV[0] || -B $ARGV[0]; $year = (localtime)[5] + 1900 }
+            s/^([*\/#\/"*[:space:]]*)Copyright\s+(?:\(C\)\s*)?(\d+)(?:\s*-\s*\d+)?\s(Advanced\s*Micro\s*Devices)/qq($1Copyright (C) $2@{[$year != $2 ? "-$year" : ""]} $3)/ie
+            if $. < 10' "$file" && git add -u "$file"
```
Comment (pfultz2) [inline .githooks/pre-commit:56]: Will this add the files to the current commit? If copyrights are updated we want to have that as a seperate commit.

File: license_stamper.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                    openAndWriteFile(file, message, commentDelim)
+
+        else:
+            print("error parsing in list of files")
+    else:
+        print("error parsing in license file")
+
+
+if __name__ == "__main__":
+    main()
```
Comment (pfultz2) [inline license_stamper.py:128]: This script should go under the tools directory.

File: license_stamper.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+def main():
+    #figure out what file we want to read in our message from
+    #    liscence_file = input("Enter License file (default is LICENSE)")
+
+    #if liscence_file is None:
+    #        licence_file = "LICENSE"
+
+    bashCommand = "cat LICENSE"
+    process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)
```
Comment (pfultz2) [inline license_stamper.py:104]: I think `open('LICENSE').read()` would be simpler.

File: license_stamper.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    message, error = process.communicate()
+
+    if error == None:
+        #Get a list of all the files in our git repo
+        bashCommand = "git ls-files --exclude-standard"
+        process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)
```
Comment (pfultz2) [inline license_stamper.py:110]: Use `subprocess.run(bashCommand, capture_output=True, check=True, shell=True)` instead.

File: examples/nlp/python_bert_squad/run_onnx_squad.py [hunk 6]
```diff
@@ -1,3 +1,26 @@
...
+#  all copies or substantial portions of the Software.
+#  
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+###########################################
```
Comment (pfultz2) [inline examples/nlp/python_bert_squad/run_onnx_squad.py:23]: The copyright below should be removed.

**Comment by pfultz2** [comment]
We should have a script so we can run a check in the CI to ensure there is a license and the copyright is up to date based on the last time it was modified.

**Comment by umangyadav** [comment]
`cpp_mnist/digit/txt` is binary file, It got caught with the script. Can you remove change for that ?

File: src/include/migraphx/half.hpp [hunk 7]
```diff
@@ -1,3 +1,26 @@
...
+*  
+*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+*  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+*  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+*  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+*  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+*  THE SOFTWARE.
+*/
 /*=============================================================================
```
Comment (umangyadav) [inline src/include/migraphx/half.hpp:24]: This one also has dual licenses.

File: src/py/backend/backend.py [hunk 8]
```diff
@@ -1,3 +1,26 @@
...
+#  
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+###########################################
 # -------------------------------------------------------------------------
```
Comment (umangyadav) [inline src/py/backend/backend.py:24]: Dual Licenses

File: src/targets/gpu/device/include/migraphx/gpu/device/types.hpp [hunk 9]
```diff
@@ -1,3 +1,26 @@
...
+*  
+*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+*  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+*  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+*  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+*  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+*  THE SOFTWARE.
+*/
 /*=============================================================================
```
Comment (umangyadav) [inline src/targets/gpu/device/include/migraphx/gpu/device/types.hpp:24]: Dual Licenses

File: doc/src/conf.py [hunk 10]
```diff
@@ -1,3 +1,26 @@
...
+#  all copies or substantial portions of the Software.
+#  
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+###########################################
```
Comment (umangyadav) [inline doc/src/conf.py:23]: Looks like this file is auto-generated.  I am not sure if we should add a license on this or not.

File: .githooks/pre-commit (hunk 1 above)
Comment (TedThemistokleous) [reply .githooks/pre-commit:49]: Should be all of them. I grabbed this from RocmBLAS commit Umang showed me.

File: doc/src/conf.py (hunk 10 above)
Comment (TedThemistokleous) [reply doc/src/conf.py:23]: How do we want to handle auto gen'd files then? This sounds liek I should have another file with a list of thing not to stamp then

File: .githooks/pre-commit (hunk 1 above)
Comment (pfultz2) [reply .githooks/pre-commit:49]: We should reuse the query from files from below at line 61 and then assign it to the `files` variable: `files=$(git diff-index --cached --name-only $against | grep -E '\.h$|\.hpp$|\.cpp$|\.cl$|\.h\.in$|\.hpp\.in$|\.cpp\.in$|\.py$')`, and then use the `$files` variable below instead.

File: doc/src/conf.py (hunk 10 above)
Comment (pfultz2) [reply doc/src/conf.py:23]: Its not completely autogenerated. It was initially generated but then we modified with out settings, It should still have a license.

Comment (TedThemistokleous) [reply doc/src/conf.py:23]: What about rules for generated files then? Can we assume since "we" generated it, we should stamp it with our license?

**Comment by TedThemistokleous** [comment]
> @TedThemistokleous #1183 got merged in which added a few extra files. When you merge/rebase with develop branch, make sure to add licenses on those files.

Do you have a list of which files to update?

**Comment by TedThemistokleous** [comment]
Updated this with changes to the script and removed the githook for now I borrowed from rocmBLAS.

File: src/include/migraphx/half.hpp [hunk 11]
```diff
@@ -1,9 +1,26 @@
-/*=============================================================================
-    Copyright (c) 2017 Paul Fultz II
-    half.hpp
-    Distributed under the Boost Software License, Version 1.0. (See accompanying
```
Comment (causten) [inline src/include/migraphx/half.hpp:4]: License changed from Boost to MIT?

File: tools/roctx.py [hunk 12]
```diff
@@ -1,3 +1,26 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#####################################################################################
 #!/usr/bin/env python3
```
Comment (causten) [inline tools/roctx.py:24]: License should come after this line

File: src/include/migraphx/type_traits.hpp [hunk 13]
```diff
@@ -1,9 +1,26 @@
-/*=============================================================================
-    Copyright (c) 2017 Paul Fultz II
-    type_traits.hpp
-    Distributed under the Boost Software License, Version 1.0. (See accompanying
-    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
-==============================================================================*/
```
Comment (causten) [inline src/include/migraphx/type_traits.hpp:6]: License changed?

File: src/targets/gpu/device/include/migraphx/gpu/device/types.hpp [hunk 14]
```diff
@@ -1,9 +1,26 @@
-/*=============================================================================
-    Copyright (c) 2017 Paul Fultz II
-    types.hpp
-    Distributed under the Boost Software License, Version 1.0. (See accompanying
-    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
-==============================================================================*/
```
Comment (causten) [inline src/targets/gpu/device/include/migraphx/gpu/device/types.hpp:6]: License changed?

**Review by causten — CHANGES_REQUESTED** [review]
I'd like to get some clarity on the files that were previously licensed under Boost.

File: tools/roctx.py (hunk 12 above)
Comment (TedThemistokleous) [reply tools/roctx.py:24]: Easy fix.

**Comment by TedThemistokleous** [comment]
> I'd like to get some clarity on the files that were previously licensed under Boost.

Doing a grep search off develop shows those three files are licensed. 

Looks like it was some files which were originally licensed by @pfultz2 

`tthemist@ixt-rack-186 ~/AMDMIGraphX (develop) $ grep -HnIR " Software License" * --exclude-dir=depend
src/include/migraphx/half.hpp:4:    Distributed under the Boost Software License, Version 1.0. (See accompanying
src/include/migraphx/type_traits.hpp:4:    Distributed under the Boost Software License, Version 1.0. (See accompanying
src/targets/gpu/device/include/migraphx/gpu/device/types.hpp:4:    Distributed under the Boost Software License, Version 1.0. (See accompanying`

**Comment by TedThemistokleous** [comment]
Looks like from the email license was erroneously added. 

I'll regenerate and revert things with an updated script that correctly handles shebangs (#!) at the start of scripts too.

@causten do we want this to also stamp our bash files as well or does the current bucket of cpp/python suffice?

File: src/include/migraphx/type_traits.hpp (hunk 13 above)
Comment (causten) [reply src/include/migraphx/type_traits.hpp:6]: > Looks like from the email license was erroneously added.
> 
> I'll regenerate and revert things with an updated script that correctly handles shebangs (#!) at the start of scripts too.
> 
> @causten do we want this to also stamp our bash files as well or does the current bucket of cpp/python suffice?

yes that resolves my comment.  Thanks for investigating!

**Comment by causten** [comment]
> Looks like from the email license was erroneously added.
> 
> I'll regenerate and revert things with an updated script that correctly handles shebangs (#!) at the start of scripts too.
> 
> @causten do we want this to also stamp our bash files as well or does the current bucket of cpp/python suffice?

All source code files need a  license.  Yeah .bsh/.sh files too

**Review by causten — APPROVED** [review]

**Comment by TedThemistokleous** [comment]
Adding in a check_stamped.py script that goes over everything in the repo and determines if it needs to be stamped + determines if there's things outstanding/we haven't accounted her.

@pfultz2 do we want to be also handling json, and .ipynb files as well? The rest of the things seem to be just git scripts, build time files, and binarys (.proto/.onnx/.doocker)

For clarity here's what I ahve in the script right now


`    # The filetypes we want to check for that are stamped
    # LICENSE is included here as it SHOULD have a liscence in it otherwise flag it as unstamped
    supported_file_types = (".cpp", ".hpp", ".h", ".py", ".txt", ".sh", ".bsh",
                            "LICENSE")

    #add general stuff we shouldn't stamp and any exceptions here
    unsupported_file_types = [
        ".onnx", ".pb", ".rst", ".jpg", ".jpeg", ".proto", ".md", ".clang",
        ".cmake", ".weight", ".ipynb", ".ini", "json", ".docker", ".git",
        ".rules", ".yml"
    ]

    #Specifics to ignore
    specificIgnores = ("digits.txt", "Dockerfile", "Jenkinsfile", "")`

**Comment by pfultz2** [comment]
> do we want to be also handling json, and .ipynb files as well?

I am not sure, thats a good question for our legal department, but I think we can start with extensions you have now and extend it in the future.

File: src/py/backend/backend_rep.py [hunk 15]
```diff
@@ -1,3 +1,26 @@
...
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+###########################################
 # -------------------------------------------------------------------------
 # Copyright (c) Advanced Micro Device Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (umangyadav) [reply src/py/backend/backend_rep.py:25]: Can you remove bottom license here ?

**Comment by causten** [comment]
> > do we want to be also handling json, and .ipynb files as well?
> 
> I am not sure, thats a good question for our legal department, but I think we can start with extensions you have now and extend it in the future.

JSON no, Jupyter Notebook's yes.

**Comment by umangyadav** [comment]
> @TedThemistokleous #1183 got merged in which added a few extra files. When you merge/rebase with develop branch, make sure to add licenses on those files.

Can you rebase wrt develop and add those files ? You can take a look at that PR to find out new files that needs to be stamped.

**Comment by TedThemistokleous** [comment]
> > @TedThemistokleous #1183 got merged in which added a few extra files. When you merge/rebase with develop branch, make sure to add licenses on those files.
> 
> Can you rebase wrt develop and add licenses in those files ? You can take a look at that PR to find out new files that needs to be stamped.

The latest script I uploaded would flag those files actually (tools/check_stampled.py)
I added a list of file types we support and don't support + some ignores incase we have some oddball exceptions (like digits.txt) . I can add those files once they get flagged as unstamped.

@causten okay, let me fix and change things for the notebook files. 

Take a look at tools/check_stamped.py which just grabs the file types we support

**Comment by TedThemistokleous** [comment]
> > > do we want to be also handling json, and .ipynb files as well?
> > 
> > 
> > I am not sure, thats a good question for our legal department, but I think we can start with extensions you have now and extend it in the future.
> 
> JSON no, Jupyter Notebook's yes.

Just added proper support for our notebook files and updated the ones in the repo.

File: tools/check_stamped.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        print("Error: The following " + str(len(unknownFiles)) +
+              " files not handled:")
+        print(str(unknownFiles))
+        sys.exit(2)
+
+    sys.exit(0)
+
+
+if __name__ == "__main__":
+    main()
```
[the remark this answers is not in the document]
Comment (TedThemistokleous) [reply tools/check_stamped.py]: https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1255

**Comment by TedThemistokleous** [comment]
Rebased off current develop branch. Targeting the other PR with the CI check and check_script to this branch so we can capture everything once this gets in.

**Pull request merged**